README



山寨版的神州数码802.1x认证supplicant，基于pcap库的C语言的跨平台的原生客户端。
原项目地址：https://code.google.com/p/zdcclient/
原开发者：Pentie、michael8090、zhhtc200、zhouJFu

由于原开发者现已离校，由Pentie授权fork至github，继续维护

changelog
1.2更新
-从googlecode转移到github
-添加了新的版本号（佛山科学技术学院样本
-可以解析及回应新版神州数码的存活包（佛山科学技术学院样本

1.1更新
-增加显示服务器返回的信息
-改善运行脚本，可通过notify显示服务器返回的信息
-增加安装脚本，简化用户的安装操作

0.9-1更新
-调整程序代码位置以及完善进程检测等；
-修正0.9发布的版本存在严重bug导致无法认证；

0.8更新
-改变程序检查运行副本、后台运行的方式

0.7更新
-加入程序独立检查，当系统中有同名进程存在时，停止执行并提示原进程pid
-加入退出进程的选项-l，可通知原进程下线并退出。

0.6更新
-修改eap头的eap_id类型，以及之前的误用：
	eap_id只作为是REQ/IDY还是REQ/IDY-KEEP ALIVE的判断；
	eap_id还用在MD5-Challenge的计算中，虽然神州数码中计算MD5-Challenge的时候eap_id总是0x02，不过还是作为参数传入，更符合RFC 1994关于EAP的标准；

0.5更新
-改善DHCP模式下的认证IP处理方法：
    若程序能够从网卡获得IP，则使用网卡的真实IP
    只有获取不到IP时才使用伪IP 169.254.216.45

-在使用伪IP的情况下，当接受到第一个REQUEST IDT KEEP
ALIVE报文后，尝试再次获取网卡IP与MASK并写入RESPONSE数据帧缓冲区替换伪IP，之后发出的RESPONSE数据帧将包含真实IP
-只有当用户同时指定了IP和MASK才使用用户设定值，否则由程序处理IP
-修正一些stderr、stderr的滥用
-添加打包脚本

0.4更新
-修改DHCP模式下的地址模式的地址，可默认提供伪IP认证；
-增加--ip、--mask参数，可自定义dhcp模式下的发送的网卡ip和掩码
-静态模式下（或在认证前就能通过DHCP）的网卡ip通过api自动获取
-DNS、GATEWAY服务器可留空
-加入分别针对DHCP模式的启动脚本和静态模式的启动脚本；

0.3更新
-客户端默认版本号更新为3.5.04.1013fk
-修正对SUCCESS数据包的识别
-后台运行改为在认证成功后主程序返回
-根据状态区分用户名、用户配置信息错误与密码错误的两种提示
-在接收和发送Keep Alive报文时提示当前线程的pid
-改善pcap的抓包过滤器，只抓发往本机的报文

0.2 更新
-整理代码
-进一步完善包解析、发送过程细节，加入遇到为止包时的提示
-区分EAP_RESPONSE_IDENTITY和EAP_RESPONSE_IDENTITY_KEEP_ALIVE发送的包
-加入自定义客户端版本号的功能参数--ver，并能校验版本号长度
-默认版本号改成3.5.04.0324
-修改--help内About ZDClient的文字
-在认证是提示协议版本信息
-增加可检测root权限的启动脚本


main.c
/*
 * =====================================================================================
 *
 *       Filename:  main.c
 *
 *    Description:  main control for zdclient.
 *
 *        Version:  1.0
 *        Created:  07/06/2009 03:53:22 PM
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  BOYPT (PT), pentie@gmail.com
 *        Company:  http://apt-blog.co.cc
 *
 * =====================================================================================
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <fcntl.h>
#include "zdclient.h"

#define LOCKFILE "/var/run/zdclient.pid"        /* 锁文件 */

#define LOCKMODE (S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)

static void signal_interrupted (int signo);

extern pcap_t      *handle;
extern int          exit_flag;

int                 lockfile;                  /* 锁文件的描述字 */

void
flock_reg ()
{
    char buf[16];
    struct flock fl;
    fl.l_start = 0;
    fl.l_whence = SEEK_SET;
    fl.l_len = 0;
    fl.l_type = F_WRLCK;
    fl.l_pid = getpid();
 
    //阻塞式的加锁
    if (fcntl (lockfile, F_SETLKW, &fl) < 0){
        perror ("fcntl_reg");
        exit(1);
    }
 
    //把pid写入锁文件
    assert (0 == ftruncate (lockfile, 0) );    
    sprintf (buf, "%ld", (long)getpid());
    assert (-1 != write (lockfile, buf, strlen(buf) + 1));
}


void
daemon_init(void)
{
	pid_t	pid;
    int     fd0;

	if ( (pid = fork()) < 0)
	    perror ("Fork");
	else if (pid != 0) {
        fprintf(stdout, "&&Info: ZDClient Forked background with PID: [%d]\n\n", pid);
		exit(0);
    }
	setsid();		/* become session leader */
	assert (0 == chdir("/tmp"));		/* change working directory */
	umask(0);		/* clear our file mode creation mask */
    flock_reg ();

    fd0 = open ("/dev/null", O_RDWR);
    dup2 (fd0, STDIN_FILENO);
    dup2 (fd0, STDERR_FILENO);
    dup2 (fd0, STDOUT_FILENO);
    close (fd0);
}


int 
program_running_check()
{
    struct flock fl;
    fl.l_start = 0;
    fl.l_whence = SEEK_SET;
    fl.l_len = 0;
    fl.l_type = F_WRLCK;
 
    //尝试获得文件锁
    if (fcntl (lockfile, F_GETLK, &fl) < 0){
        perror ("fcntl_get");
        exit(1);
    }

    if (exit_flag) {
        if (fl.l_type != F_UNLCK) {
            if ( kill (fl.l_pid, SIGINT) == -1 )
                perror("kill");
            fprintf (stdout, "&&Info: Kill Signal Sent to PID %d.\n", fl.l_pid);
        }
        else 
            fprintf (stderr, "&&Info: NO ZDClient Running.\n");
        exit (EXIT_FAILURE);
    }

    //没有锁，则给文件加锁，否则返回锁着文件的进程pid
    if (fl.l_type == F_UNLCK) {
        flock_reg ();
        return 0;
    }

    return fl.l_pid;
}

static void
signal_interrupted (int signo)
{
    fprintf(stdout,"\n&&Info: USER Interrupted. \n");
    send_eap_packet(EAPOL_LOGOFF);
    pcap_breakloop (handle);
}


int main(int argc, char **argv)
{
    //初始化并解释程序的启动参数
    init_arguments (&argc, &argv);

    //打开锁文件
    lockfile = open (LOCKFILE, O_RDWR | O_CREAT , LOCKMODE);
    if (lockfile < 0){
        perror ("Lockfile");
        exit(1);
    }

    //检测程序的副本运行（文件锁）
    int ins_pid;
    if ( (ins_pid = program_running_check ()) ) {
        fprintf(stderr,"@@ERROR: ZDClient Already "
                            "Running with PID %d\n", ins_pid);
        exit(EXIT_FAILURE);
    }

    //初始化用户信息
    init_info();

    //初始化设备，打开网卡，获得Mac、IP等信息
    init_device();

    //初始化发送帧的缓冲区
    init_frames ();

    signal (SIGINT, signal_interrupted);
    signal (SIGTERM, signal_interrupted);    
    show_local_info();

    //发出第一个上线请求报文
    send_eap_packet (EAPOL_START);

    //进入回呼循环。以后的动作由回呼函数get_packet驱动，
    //直到pcap_break_loop执行，退出程序。
	pcap_loop (handle, -2, get_packet, NULL);   /* main loop */
    pcap_close (handle);
    return 0;
}






zdclient.h
/*
 * =====================================================================================
 *
 *       Filename:  zdclient.h
 *
 *    Description:
 *
 *        Version:  1.0
 *        Created:  06/06/2009 03:47:25 PM
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  YOUR NAME (),
 *        Company:
 *
 * =====================================================================================
 */

#include <pcap.h>
#include <stdio.h>
#include <stdlib.h>

#include <string.h>
#include <ctype.h>
#include <errno.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/stat.h>

#include <netinet/in.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <net/ethernet.h>


#ifndef __linux
//------bsd/apple mac
    #include <net/if_var.h>
    #include <net/if_dl.h>
    #include <net/if_types.h>
#endif

#include <getopt.h>
#include <iconv.h>
#include "md5.h"

/* ZDClient Version */
#define ZDC_VER "0.13"

/* default snap length (maximum bytes per packet to capture) */
#define SNAP_LEN 1518

/* ethernet headers are always exactly 14 bytes [1] */
#define SIZE_ETHERNET 14

struct eap_header {
    uint8_t eapol_v;
    uint8_t eapol_t;
    uint16_t eapol_length;
    uint8_t eap_t;
    uint8_t eap_id;
    uint16_t eap_length;
    uint8_t eap_op;
    uint8_t eap_v_length;
    uint8_t eap_md5_challenge[16];
};

struct dcba_tailer {
    bpf_u_int32     local_ip;
    bpf_u_int32     local_mask;
    bpf_u_int32     local_gateway;
    bpf_u_int32     local_dns;
    uint8_t          username_md5[16];
    uint8_t          client_ver[13];
};

enum EAPType {
    EAPOL_START,
    EAPOL_LOGOFF,
    EAP_REQUEST_IDENTITY,
    EAP_RESPONSE_IDENTITY,
    EAP_REQUEST_IDENTITY_KEEP_ALIVE,
    EAP_RESPONSE_IDENTITY_KEEP_ALIVE,
    EAP_REQUETS_MD5_CHALLENGE,
    EAP_RESPONSE_MD5_CHALLENGE,
    EAP_SUCCESS,
    EAP_FAILURE,
    ERROR,
    EAP_REQUEST_MD5_KEEP_ALIVE=250
};

enum STATE {
   READY,
   STARTED,
   ID_AUTHED,
   ONLINE
};

void    send_eap_packet(enum EAPType send_type);
void    show_usage();
char*   get_md5_digest(const char* str, size_t len);
void    action_by_eap_type(enum EAPType pType,
                        const struct eap_header *header,
                        const struct pcap_pkthdr *packetinfo,
                        const uint8_t *packet);
void    send_eap_packet(enum EAPType send_type);
void    init_frames();
void    init_info();
void    init_device();
void    init_arguments(int *argc, char ***argv);
int     set_device_new_ip();
void    fill_password_md5(uint8_t attach_key[], uint8_t eap_id);
void    fill_uname_md5(uint8_t attach_key[], uint8_t eap_id);
int     program_running_check();
void    daemon_init(void);
void    show_local_info();
void    print_server_info (const uint8_t *packet, uint16_t packetlength);
int     code_convert(char *from_charset, char *to_charset,
             char *inbuf, size_t inlen, char *outbuf, size_t outlen);


void
get_packet(uint8_t *args, const struct pcap_pkthdr *header,
    const uint8_t *packet);


